### PR TITLE
add first tests

### PR DESCRIPTION
**Pull request by slaroche**

### CODE REVIEW

**Comment by aviau** [comment]
- Can you add a target in the makefile to run the tests?
- Can you configure github actions to run the tests?
- Can you configure github actions to check for go fmt too?

File: pkg/api/api_test.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+		t.Fatal(err)
+	}
+
+}
+
+func TestQuery(t *testing.T) {
+	ts := httptest.NewServer(NewRouter())
+	defer ts.Close()
+
+	_, err := http.Get(strings.Join([]string{ts.URL, "api/query"}, "/"))
```
Comment (aviau) [inline pkg/api/api_test.go]: Just so you know, you can isolate err to a scope and do this in one block like this:
```go
if _, err := http.Get(strings.Join([]string{ts.URL, "api/query"}, "/")); err != nil {
    t.Fatal(err)
}
```

File: pkg/api/api_test.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
+package api
```
Comment (aviau) [inline pkg/api/api_test.go]: Can we move this to api_test?

This makes it clearer that we are testing the package's public API

**Comment by aviau** [comment]
Can we create a function that setups the API tests so that we can reuse it in future tests?

We already have two tests that setutp the same boilerplate.

Right now creating the test server is very simple but it won't stay like that for long so we may as well start encapsulating it.

```go

type ApiTest struct {
   server *httptest.Server
    t    *testing.T
}

func (apiTest *ApiTest) Close() {
    apiTest.server.Close()
}

func newApiTest(t *testing.T) *ApiTest {
    apiTest := ApiTest{
        server: httptest.NewServer(NewRouter())
    }
    return &apiTest
}


func TestApi(t *testing.T) {
        apiTest := NewApiTest()
	defer apiTest.Close()

	_, err := http.Get(strings.Join([]string{ts.URL, "api/status"}, "/"))
	if err != nil {
		t.Fatal(err)
	}

}

```

**Comment by aviau** [comment]
For github actions feel free to copy my setup here:
- https://github.com/perkeep/perkeep/blob/master/.github/workflows/tests-linux.yml

File: pkg/api/api_test.go (hunk 2 above)
Comment (slaroche) [reply pkg/api/api_test.go]: should I move it in `pkg/api_test` or `pkg/api/api_test`?

Comment (aviau) [reply pkg/api/api_test.go]: Don't move the file.

Just rename the package. `package api_test`

golang knows not to compile test packages when building programs.

Comment (slaroche) [reply pkg/api/api_test.go]: And I though that the idea to have api_test.go in the same filte and package of api.go was to have access to all funcs!?

Comment (aviau) [reply pkg/api/api_test.go]: You are only using public funcs.

**Comment by slaroche** [comment]
ready for second round of review!

@aviau

File: pkg/loki/client.go [hunk 3]
```diff
@@ -8,12 +8,14 @@ import (
 )
 
 type LokiClient struct {
-	baseUrl string
+	Client  *http.Client
+	BaseUrl string
```
Comment (aviau) [inline pkg/loki/client.go]: does BaseUrl and client need to be exported?

Comment (slaroche) [reply pkg/loki/client.go]: for the tests yes I think so.

Comment (aviau) [reply pkg/loki/client.go]: but I don't see them used in tests?

File: pkg/api/api_test.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+import (
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"testing"
+
+	"github.com/flared/lokify/pkg/api"
+)
+
+type ApiTest struct {
```
Comment (aviau) [inline pkg/api/api_test.go]:
```
apiTest
```

File: pkg/loki/client.go [hunk 5]
```diff
@@ -8,12 +8,14 @@ import (
 )
 
 type LokiClient struct {
```
Comment (aviau) [inline pkg/loki/client.go]: ```lokiClient```

File: pkg/loki/client_test.go [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+			ResultType: "streams",
+			Result: []loki.QueryDataResult{
+				{
+					Stream: map[string]string{"label": "value"},
+					Values: [][2]string{{"123456789", "log message"}},
+				},
+			},
+		},
+	}
+	respBodyBytes := new(bytes.Buffer)
```
Comment (aviau) [inline pkg/loki/client_test.go]:
```
var respBodyBytes bytes.Buffer
```

File: pkg/loki/client_test.go [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+	json.NewEncoder(respBodyBytes).Encode(respBody)
```
Comment (aviau) [inline pkg/loki/client_test.go]: ignored error
```go
if err := json.NewEncoder(respBodyBytes).Encode(respBody); err != nil {
    t.Fatal(err)
}
```

File: pkg/loki/client_test.go [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+		expectedUrl := "/loki/api/v1/query?limit=100&query=%7Blabel%3D%22value%22%7D"
+		if req.URL.String() != expectedUrl {
+			t.Fatalf("got %v, expect %v", req.URL.String(), expectedUrl)
+		}
+		rw.Write(respBodyBytes.Bytes())
+	}))
+	defer server.Close()
+
+	api := loki.LokiClient{server.Client(), server.URL}
+	resp, err := api.Query(`{label="value"}`)
```
Comment (aviau) [inline pkg/loki/client_test.go]:
```go
if resp, err := api.Query(`{label="value"}`); err != nil {
    t.Fatal(err)
} else if !reflect.DeepEqual(resp, respBody) {
    t.fatal(..)...
}
```

File: pkg/loki/client_test.go (hunk 6 above)
Comment (slaroche) [reply pkg/loki/client_test.go]: json.NewEncoder(buffer) doesn't seem to accept nil

```
	var respBodyBytes bytes.Buffer
	if err := json.NewEncoder(respBodyBytes).Encode(respBody); err != nil {
		t.Fatal(err)
	}

output:
var respBodyBytes bytes.Buffer
cannot use respBodyBytes (variable of type bytes.Buffer) as io.Writer value 
in argument to json.NewEncoder: missing method Write 
(Write has pointer receiver) compilerInvalidIfaceAssign)
```

Comment (aviau) [reply pkg/loki/client_test.go]: hmm maybe just `b := &bytes.Buffer{}`

Comment (aviau) [reply pkg/loki/client_test.go]: or `var b *bytes.Buffer`

Comment (slaroche) [reply pkg/loki/client_test.go]: `&bytes.Buffer{}` is better then `new(bytes.Buffer)` ?

Comment (slaroche) [reply pkg/loki/client_test.go]: ok you are right !
this is working:
```
var b *bytes.Buffer
```

Thanks!

Comment (slaroche) [reply pkg/loki/client_test.go]: oh never mind

```
--- FAIL: TestQuery (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x4e70e6]
```

**Pull request merged**